### ISSUE TITLE
Different results for add_loss() within TimeDistributed for eager vs. compiled execution of keras Layer

### ISSUE THREAD

**Issue by djoshea** [tensorflow/tensorflow#46244]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes, included below
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): pip binary
- TensorFlow version (use command below): v2.4.0-rc4-71-g582c8d236cb 2.4.0
- Python version: '3.8.6 | packaged by conda-forge | (default, Nov 27 2020, 19:31:52) \n[GCC 9.3.0]'
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: 11.0 / 8
- GPU model and memory: TITAN X (Pascal) computeCapability: 6.1

**Describe the current behavior**

In the minimal example code below, I have a keras Layer which calls `add_loss()`, and this layer is wrapped in `TimeDistributed`. When I execute the layer in eager mode, the losses have an extra timestep, as the first timestep is double-counted. However, if I evaluate the compiled model, the correct loss is computed.

During eager execution, the `add_loss` gets called first when computing `output_timestep_zero`:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/backend.py#L4452

and subsequently called 5 (== # timesteps) times by: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/backend.py#L4538

**Describe the expected behavior**

I expected layer.losses to count each timestep once and not double count the first timestep. I would think that the loss would be cleared after the first call used to obtain `output_timestep_zero`. I'm not entirely sure I understand why the first timestep is double-counted, but it would be useful to have the eager mode execution return the same losses as reported when the model is compiled.

**Standalone code to reproduce the issue**

below I run a simple example with 5 timesteps, where the input equals the timestep value + 1, so the correct loss would be
`mean([1, 2, 3, 4, 5]) == 3`. In eager execution, the first timestep is counted twice

```python
import tensorflow as tf
import tensorflow.keras as keras
import numpy as np

# build inputs(b, t, i) == t + 1 with size [batch, timesteps, inputs]
batch, timesteps, ninputs = 1, 5, 1
timevec = (
    np.arange(timesteps, dtype=np.float32) + 1
)  # we start at one since the first gets double-counted
inputs = np.broadcast_to(
    timevec[np.newaxis, :, np.newaxis], (batch, timesteps, ninputs)
)


class AddsLossLayer(keras.layers.Layer):
    """Identity layer which calls add_loss on mean of input"""

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def call(self, inputs):
        # inputs is batch x ninputs
        # return input but add_loss(mean of current input over ninputs dimension, should == timestep + 1)
        self.add_loss(tf.math.reduce_mean(inputs, axis=0, name="loss"))
        return inputs

    def compute_output_shape(self, input_shape):
        return input_shape


adds_loss_layer = AddsLossLayer()
td_layer = keras.layers.TimeDistributed(adds_loss_layer)

outputs = td_layer(inputs)

print(f"inputs shape = {inputs.shape}")
print(f"outputs shape = {outputs.shape}")
print(f"len(td_layer.losses) == {len(td_layer.losses)}")
print(f"len(adds_loss_layer.losses) == {len(adds_loss_layer.losses)}")
print("loss values == ", [loss.numpy() for loss in td_layer.losses])

# inputs shape = (3, 5, 1)
# outputs shape = (3, 5, 1)
# len(td_layer.losses) == 6
# len(adds_loss_layer.losses) == 6
# loss values ==  [array([1.], dtype=float32), array([1.], dtype=float32), array([2.], dtype=float32), array([3.], dtype=float32), array([4.], dtype=float32), array([5.], dtype=float32)]

# so we double count the first timestep!

model_inputs = tf.keras.Input(shape=inputs.shape[1:], dtype="float32")
adds_loss_layer = AddsLossLayer()
td_layer = keras.layers.TimeDistributed(adds_loss_layer, name="td")
model_outputs = td_layer(model_inputs)

model = keras.Model(model_inputs, model_outputs)
model.compile()
model.evaluate(inputs, None, verbose=2)

# 1/1 - 0s - loss: 3.0000
# (this is correct, (1 + 2 + 3 + 4 + 5) / 5 == 3)

model.fit(inputs, None, epochs=1)

# 1/1 [==============================] - 0s 100ms/step - loss: 3.0000
# (this is correct, (1 + 2 + 3 + 4 + 5) / 5 == 3)

outputs = model(inputs)

print("mean model.losses = ", tf.math.reduce_mean(model.losses).numpy())

# mean model.losses =  2.6666667
# due to double counted first timestep, (1 + 1 + 2 + 3 + 4 + 5) / 6 = 2.666667
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3, TF v2.4 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/b22c6930b5dc31c750993ee0de021c8b/46244.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
@djoshea I can reproduce this with recent `TF2.6` and `tf-nightly`. May be this is a potential bug. 

As Keras moved to separate repo (i can't move the issue), can you please close here and open in keras repo https://github.com/keras-team/keras/issues

Thanks!
